packages/nimble/inst/tests/test_utils.R

```diff
@@ -26,6 +26,49 @@ temporarilyAssignInGlobalEnv <- function(value) {
     do.call('on.exit', list(rmCommand), envir = parent.frame())
 }
 
+withTempProject <- function(code) {
+    code <- substitute(code)
+    project <- nimble:::nimbleProjectClass()
+    nimbleOptions(nimbleProject = project)
+    on.exit({
+        ## messages are suppressed by test_that, so "assign('.check', 1, globalenv())" can be used as a way to verify this code was called 
+        .project <- nimbleOptions()$nimbleProject
+        nimbleOptions(nimbleProject = NULL) ## clear this before clearCompiled step, in case clearCompiled() itself fails
+        .project$clearCompiled()
+    })
+    try(eval(code, envir = parent.frame()))
```

Comment:
Using `tryCatch` did not seem meaningfully different to me.  However I note `?on.exit` indicates that only one `on.exit` instruction will be stored for a function.  We may be at risk of `temporarilyAssignInGlobalNamespace` and `withTempProject` stepping on each other's use of `on.exit.